src/api/spec/controllers/webui/download_on_demand_controller_spec.rb

```diff
@@ -0,0 +1,165 @@
+        repotype:             "rpmmd",
+        url:                  "http://mola.org",
+        archfilter:           "i586",
+        masterurl:            "http://opensuse.org",
+        mastersslfingerprint: "asdfasd",
+        pubkey:               "3jnlkdsjfoisdjf0932juro2ikjfdslñkfj",
+        repository_id:        repository.id
+      }
+    }
+  }
+
+  before do
+    project.repositories << repository
+  end
+
+  it "uses strong parameters" do
+    login(admin_user)
+    should permit(:arch, :repotype, :url, :repository_id, :archfilter, :masterurl, :mastersslfingerprint, :pubkey).
+              for(:create, params: dod_parameters).on(:download_repository)
+  end
+
+  describe "POST create" do
+    it { is_expected.to use_before_action(:set_project) }
+
+    context "permission check" do
+      skip("Ensure that users can't add dod repositories to repos / projects they have no write access to!")
+    end
+
+    context "for non-admin users" do
+      before do
+        login(create(:confirmed_user))
+        post :create, dod_parameters
+      end
+
+      it { is_expected.to redirect_to(root_path) }
+      it { expect(flash[:error]).to eq("Sorry, you are not authorized to create this DownloadRepository.") }
+      it { expect(DownloadRepository.where(dod_parameters[:download_repository])).not_to exist }
+    end
+
+    context "valid requests" do
+      before do
+        login(admin_user)
+        post :create, dod_parameters
+      end
+
+      it { is_expected.to redirect_to(project_repositories_path(project)) }
+      it { expect(flash[:notice]).to eq("Successfully created Download on Demand") }
+      it { expect(assigns(:download_on_demand)).to be_kind_of(DownloadRepository) }
+      it { expect(DownloadRepository.where(dod_parameters[:download_repository])).to exist }
+    end
+
+    context "invalid parameter (architecture)" do
+      before do
+        dod_parameters[:download_repository][:arch] = ""
+        login(admin_user)
+        post :create, dod_parameters
+      end
+
+      it { is_expected.to redirect_to(root_path) }
+      it { expect(flash[:error]).to eq("Download on Demand can't be created: Arch can't be blank") }
+      it { expect(assigns(:download_on_demand)).to be_kind_of(DownloadRepository) }
+      it { expect(DownloadRepository.where(dod_parameters[:download_repository])).not_to exist }
+    end
+  end
+
+  describe 'DELETE destroy' do
+    let(:dod_repository) { create(:download_repository) }
+
+    before do
+      repository.download_repositories << dod_repository
+    end
+
+    it { is_expected.to use_before_action(:set_project) }
+
+    context "for non-admin users" do
+      before do
+        login(create(:confirmed_user))
+        delete :destroy, id: dod_repository.id, project: project.name
+      end
+
+      it { is_expected.to redirect_to(root_path) }
+      it { expect(flash[:error]).to eq("Sorry, you are not authorized to delete this DownloadRepository.") }
```

Comment:
Such kind of before_filter checking because the admin rights as @hennevogel said is covered by the `use_before_action` call. So, no need to check it here. See the comments of henne in my PR (https://github.com/openSUSE/open-build-service/pull/1549 the outdated ones)...